library/std/src/sys/unix/fs.rs

```diff
@@ -1162,7 +1162,7 @@ pub fn copy(from: &Path, to: &Path) -> io::Result<u64> {
             };
             if let Err(ref copy_err) = copy_result {
                 match copy_err.raw_os_error() {
-                    Some(libc::ENOSYS) | Some(libc::EPERM) => {
+                    Some(libc::ENOSYS | libc::EPERM | libc::EOPNOTSUPP) => {
```

Comment:
side note `EOPNOTSUPP` and `EPERM` setting `HAS_COPY_FILE_RANGE` sound like a pessimisation to me.

I can see either of those errors occurring transiently (e.g. when copying to/from NFS or copying to a file system with no write permissions) on a system that otherwise supports this syscall just fine...

Reply:
I think `EPERM` was here because of seccomp failures noted below, where you might as well act like the syscall doesn't exist. I suppose a seccomp filter could be more nuanced than completely disabling this syscall though. However, if the filesystem had no write permission, wouldn't we have failed earlier trying to open that file handle at all?

`EOPNOTSUPP` could go either way, but I think that's mostly a quirk of RHEL7's bug. AIUI, in newer kernels VFS will fallback to splice automatically, so the user wouldn't even see it.

Reply:
[EOPNOTSUPP shouldn't be visible to userspace](https://github.com/torvalds/linux/commit/64bf5ff58dff757253cf2142542672de4b21cd1a#diff-63018726069897e4a3a3b5b03ff71613R1673), the kernel will (should) fallback to a generic splice in that case.
EPERM may indeed be a pessimization in some cases, afaict it exists to detect seccomp containers where the syscall is blocked and thus will never succeed and thus be a waste of syscalls.

If we want to change that that should be done in a separate PR imo.

Reply:
> EOPNOTSUPP shouldn't be visible to userspace

We’re already taking time to handle this case that shouldn’t ever happen, though. I was and still am a proponent of not spending any time, bytes or effort on broken setups and kernels, but the opinion held by others appears to be different. So we might as well do it right.

> However, if the filesystem had no write permission, wouldn't we have failed earlier trying to open that file handle at all?

I can still imagine cases where this could be overly pessimistic. Especially with network filesystems where having a writable file descriptor does not necessarily mean you’ll have permission to write to it (anymore).

> If we want to change that that should be done in a separate PR imo.

Fair.

Reply:
Use of `copy_file_range`  is an optimization. if you have a buggy kernel which violates the stated contract in the manpage I think it is fair game to just not apply this optimization and we don't have to jump through extra hoops to detect what kind of specific unexpected error we're getting.